packages/keycloak-service/src/service/tests/index.test.ts

```diff
@@ -0,0 +1,341 @@
+      'hunter2',
+      'users',
+      'https://keycloak-test.smartregister.org/auth/realms/'
+    );
+    let error;
+    try {
+      await usersService.create({ foo: 'bar' });
+    } catch (e) {
+      error = e;
+    }
+    expect(error).toEqual(
+      new HTTPError({}, undefined, 'KeycloakAPIService create on users failed, HTTP status 500')
+    );
+  });
+
+  it('KeycloakService update method works', async () => {
+    fetch.mockResponseOnce(JSON.stringify({}));
+    const usersService = new KeycloakService(
+      'hunter2',
+      'users',
+      'https://keycloak-test.smartregister.org/auth/realms/'
+    );
+    const obj = {
+      ...keycloakUser,
+      firstName: 'Demo1',
+    };
+    const result = await usersService.update(obj);
+    expect(result).toEqual({});
+    expect(fetch.mock.calls).toEqual([
+      [
+        'https://keycloak-test.smartregister.org/auth/realms/users',
+        {
+          'Cache-Control': 'no-cache',
+          Pragma: 'no-cache',
+          body: JSON.stringify(obj),
+          headers: {
+            accept: 'application/json',
+            authorization: 'Bearer hunter2',
+            'content-type': 'application/json;charset=UTF-8',
+          },
+          method: 'PUT',
+        },
+      ],
+    ]);
+  });
+
+  it('KeycloakService update method should handle http errors', async () => {
+    fetch.mockResponseOnce(JSON.stringify({}), { status: 500 });
+    const usersService = new KeycloakService(
+      'hunter2',
+      'users',
+      'https://keycloak-test.smartregister.org/auth/realms/'
+    );
+    let error;
+    try {
+      await usersService.update(keycloakUser);
+    } catch (e) {
+      error = e;
+    }
+    expect(error).toEqual(
+      new HTTPError({}, undefined, 'KeycloakAPIService update on users failed, HTTP status 500')
+    );
+  });
+  it('readFile works correctly', async () => {
+    const mockClass = new KeycloakService(
+      'hunter2',
+      'users',
+      'https://keycloak-test.smartregister.org/auth/realms/'
+    );
+    const mockReadFile = jest.fn();
+    mockReadFile.mockReturnValue(
+      Promise.resolve(new Blob(['greetings', 'hello'], { type: 'text/csv' }))
+    );
+    const result = await mockClass.readFile('');
+    expect(result).toEqual(expect.any(Object));
+    mockReadFile.mockRejectedValue('Error!');
+    expect(result).toEqual(expect.any(Object));
+  });
+});
+
+describe('src/errors', () => {
+  it('does not create a network error', () => {
+    /// increase test coverage.
+    try {
+      const error = new SyntaxError();
+      throwNetworkError(error);
+    } catch (err) {
+      expect(err.name).toEqual('SyntaxError');
+    }
+  });
+
+  it('creates a network error', () => {
+    /// increase test coverage.
+    try {
+      const error = new TypeError();
+      throwNetworkError(error);
+    } catch (err) {
+      expect(err.name).toEqual('NetworkError');
+    }
+  });
+
+  it('throws HTTPErrors', async () => {
```

Comment:
This test throws this warning.  Lets use act to prevent that

```
  console.error
    Warning: An update to Formik inside a test was not wrapped in act(...).
    
    When testing, code that causes React state updates should be wrapped into act(...):
    
    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */
    
    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://fb.me/react-wrap-tests-with-act
        in Formik (created by UserForm)
        in div (created by UserForm)
        in UserForm (created by WrapperComponent)
        in WrapperComponent
```

Reply:
addressed via https://github.com/OpenSRP/web/pull/39/commits/9afd8269acc9cc81d2c51c701f003e3a30c82283